libraries/common/scanner.effekt

```diff
@@ -0,0 +1,288 @@
+module scanner
+
+import char
+import stream
+
+interface Scan[A] {
+  /// Return the next character, not advancing.
+  /// That is, this does not change what any future calls on the Scanner return.
+  def peek(): A / stop
+  /// Advance the Scanner to the next character.
+  def skip(): Unit / stop
+}
+
+/// Advance the Scanner to the next character, returning it.
+def read[A](): A / { Scan[A], stop } = {
+  val t = do peek[A]()
+  do skip[A]()
+  return t
+}
+
+/// Run a scanner by reading from an input stream.
+def scanner[A, R] { scanner: () => R / Scan[A] }: R / read[A] = {
+  var last = None()
+  try {
+    scanner()
+  } with Scan[A] {
+    def peek() = resume {
+      last match {
+        case None() =>
+          val t = do read[A]()
+          last = Some(t)
+          return t
+        case Some(t) =>
+          return t
+      }
+    }
+    def skip() = resume {
+      last match {
+        case None() =>
+          val _ = do read[A]()
+          return ()
+        case Some(t) =>
+          last = None()
+          return ()
+      }
+    }
+  }
+}
+
+/// Run a scanner by reading from an input stream, discarding its output.
+def scanner[A] { scanner: () => Any / Scan[A] }: Unit / read[A] = {
+  scanner[A, Any]{scanner}
+  ()
+}
+
+
+effect fail(): Nothing
+
+def many[A] { action: () => A / fail }: Unit / emit[A] =
+  try {
+    def go(): Unit = {
+      do emit(action())
+      go()
+    }
+    go()
+  } with fail {
+    ()
+  }
+
+def successful[R] { program: () => R / fail }: Option[R] =
+  try {
+    Some(program())
+  } with fail {
+    None()
+  }
+
+def successful { program: () => Any / fail }: Bool =
+  successful[Any]{program}.isDefined
+
+def attempt[R] { program: () => R / fail } { fallback: () => R }: R =
+  try {
+    program()
+  } with fail {
+    fallback()
+  }
+
+/// Check that the next token satisfies the predicate and skip and return it when it does.
+def readIf[A] { predicate: A => Bool }: A / { Scan[A], fail, stop } = {
+  val t = do peek[A]()
+  if (predicate(t)) {
+    do skip[A]()
+    return t
+  } else {
+    do fail()
+  }
+}
+
+/// Check that the next token satisfies the predicate and skip it when it does.
+def skipIf[A] { predicate: A => Bool }: Unit / { Scan[A], fail, stop } = {
```

Comment:
Is it really necessary that `skipIf` uses `fail`? It shouldn't fail if the next one does not match, right?